### ISSUE TITLE
Unit Testing

### ISSUE THREAD

**Issue by tiferrei** [tiferrei/FAIC#3]
Test-driven development is essential for such a complex project. Most modern programming languages provide built-in test frameworks. This doesn't seem to be the case with C++. There are, however, some great open source frameworks available. The major players seem to be:

### [**Google Test**](https://github.com/google/googletest)
```cpp
// Tests factorial of 0.
TEST(FactorialTest, HandlesZeroInput) {
  EXPECT_EQ(1, Factorial(0));
}

// Tests factorial of positive numbers.
TEST(FactorialTest, HandlesPositiveInput) {
  EXPECT_EQ(1, Factorial(1));
  EXPECT_EQ(2, Factorial(2));
  EXPECT_EQ(6, Factorial(3));
  EXPECT_EQ(40320, Factorial(8));
}
```
* 👍 PROS
    * Widely-used, used by LLVM which is the base of this project.
    * Very active community, almost weekly commits.
    * Clean and easy to use 
* 👎  CONS
    * Build system has to be adapted
    * Not header only

### [**CATCH2**](https://github.com/catchorg/Catch2)
```cpp
SCENARIO( "vectors can be sized and resized", "[vector]" ) {

    GIVEN( "A vector with some items" ) {
        std::vector<int> v( 5 );
        
        REQUIRE( v.size() == 5 );
        REQUIRE( v.capacity() >= 5 );
        
        WHEN( "the size is increased" ) {
            v.resize( 10 );
            
            THEN( "the size and capacity change" ) {
                REQUIRE( v.size() == 10 );
                REQUIRE( v.capacity() >= 10 );
            }
        }
        WHEN( "the size is reduced" ) {
            v.resize( 0 );
            
            THEN( "the size changes but not capacity" ) {
                REQUIRE( v.size() == 0 );
                REQUIRE( v.capacity() >= 5 );
            }
        }
        WHEN( "more capacity is reserved" ) {
            v.reserve( 10 );
            
            THEN( "the capacity changes but not the size" ) {
                REQUIRE( v.size() == 5 );
                REQUIRE( v.capacity() >= 10 );
            }
        }
        WHEN( "less capacity is reserved" ) {
            v.reserve( 0 );
            
            THEN( "neither size nor capacity are changed" ) {
                REQUIRE( v.size() == 5 );
                REQUIRE( v.capacity() >= 5 );
            }
        }
    }
}
```
* 👍 PROS
    * Extremely easy to use.
    * Header only, minimal changes to the build system.
* 👎  CONS
    * Might be slower when used in separate binaries.

### [**Boost Test**](https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/index.html)
```cpp
#include <boost/test/included/unit_test.hpp>
using namespace boost::unit_test;

void free_test_function()
{
  BOOST_TEST( true /* test assertion */ );
}

test_suite* init_unit_test_suite( int /*argc*/, char* /*argv*/[] )
{
  if( framework::master_test_suite().argc > 1 )
    return 0;

  framework::master_test_suite().
    add( BOOST_TEST_CASE( &free_test_function ) );

  return 0;
}
```
* 👍 PROS
    * Part of Boost, carries a strong reputation.
    * We already use boost so that's in our favour.
* 👎  CONS
    * A bit older
    * Not so modern approach.

**Comment by tiferrei** [comment]
Due to the great advantages of GTest and it being used by LLVM I've decided to use it for FAIC. A PR will be submitted soon to add unit test support.

**Comment by tiferrei** [comment]
Implemented on #4
